src/napi.h

```diff
@@ -0,0 +1,663 @@
+    void Set(const char* utf8name, const char* utf8value);
+    void Set(const char* utf8name, bool boolValue);
+    void Set(const char* utf8name, double numberValue);
+
+    void DefineProperty(const PropertyDescriptor& property);
+    void DefineProperties(const std::vector<PropertyDescriptor>& properties);
+    bool InstanceOf(const Function& constructor) const;
+    template<typename T> T* Unwrap() const;
+  };
+
+  class Array : public Value {
+  public:
+    static Array New(Napi::Env env);
+    static Array New(Napi::Env env, int length);
+
+    Array();
+    Array(napi_env env, napi_value value);
+
+    Value operator [](uint32_t index) const;
+
+    uint32_t Length() const;
+    bool Has(uint32_t index) const;
+    Value Get(uint32_t index) const;
+    void Set(uint32_t index, const Value& value);
+    void Set(uint32_t index, const char* utf8value);
+    void Set(uint32_t index, bool boolValue);
+    void Set(uint32_t index, double numberValue);
+  };
+
+  class ArrayBuffer : public Value {
+  public:
+    static ArrayBuffer New(Napi::Env env, size_t byteLength);
+    static ArrayBuffer New(Napi::Env env, void* externalData, size_t byteLength);
+
+    ArrayBuffer();
+    ArrayBuffer(napi_env env, napi_value value);
+
+    void* Data();
+    size_t ByteLength();
+
+  private:
+    void* _data;
+    size_t _length;
+  };
+
+  class TypedArray : public Value {
+  public:
+    TypedArray();
+    TypedArray(napi_env env, napi_value value);
+
+    napi_typedarray_type TypedArrayType() const;
+    uint8_t ElementSize() const;
+    size_t ElementLength() const;
+    size_t ByteOffset() const;
+    size_t ByteLength() const;
+    ArrayBuffer ArrayBuffer() const;
+
+    Int8Array AsInt8Array() const;
+    Uint8Array AsUint8Array() const;
+    Uint8ClampedArray AsUint8ClampedArray() const;
+    Int16Array AsInt16Array() const;
+    Uint16Array AsUint16Array() const;
+    Int32Array AsInt32Array() const;
+    Uint32Array AsUint32Array() const;
+    Float32Array AsFloat32Array() const;
+    Float64Array AsFloat64Array() const;
+
+  protected:
+    napi_typedarray_type _type;
+    size_t _length;
+
+    TypedArray(napi_env env, napi_value value, napi_typedarray_type type, size_t length);
+  };
+
+  template <typename T, napi_typedarray_type A>
+  class _TypedArray : public TypedArray {
+  public:
+    static _TypedArray New(Napi::Env env, size_t elementLength);
+    static _TypedArray New(Napi::Env env,
+                           size_t elementLength,
+                           Napi::ArrayBuffer arrayBuffer,
+                           size_t bufferOffset);
+
+    _TypedArray();
+    _TypedArray(napi_env env, napi_value value);
+
+    T& operator [](size_t index);
+    const T& operator [](size_t index) const;
+
+    T* Data();
+    const T* Data() const;
+
+  private:
+    T* _data;
+
+    _TypedArray(napi_env env, napi_value value, size_t length, T* data);
+  };
+
+  class Function : public Value {
```

Comment:
I can see some value in `Function` inheriting from `Object`, since it is not too unusual to access properties on a function. But still without that, if you want to treat a function as an object it is trivial to call `.AsObject()` first. So, I don't feel strongly either way.